server/middleware/security/authentication.js

```diff
@@ -84,9 +84,33 @@ export const checkJwtExpiry = (req, res, next) => {
 export const _authenticateUserByJwt = (req, res, next) => {
   if (!req.jwtPayload) return next();
   const userid = Number(req.jwtPayload.sub);
+  const lastLoginAt = req.jwtPayload.lastLoginAt;
   User.findByPk(userid)
     .then(user => {
       if (!user) throw errors.Unauthorized(`User id ${userid} not found`);
+      /**
+       * Functionality for one-time login links. It checks if the JWT has
+       * lastLoginAt in the payload, which it only has during the first
+       * auth.
+       */
+      if (req.jwtPayload.lastLoginAt) {
+        /**
+         * If the user has just signed up, the JWT has a payload of 'firstLogin'
+         * and lastLoginAt is null in the db; so we allow them in only if both
+         * conditions are met. For normal logins, we check that the lastLoginAt
+         * in the JWT matches the lastLoginAt in the db. If so, we allow the user
+         * to log in, and update the lastLoginAt.
+         */
+        if (req.jwtPayload.lastLoginAt !== 'firstLogin' || user.lastLoginAt) {
+          const lastLoginInJwt = new Date(lastLoginAt);
+          const lastLoginInDb = user.lastLoginAt;
+          if (lastLoginInJwt.getTime() !== lastLoginInDb.getTime()) {
+            throw errors.Unauthorized('Invalid login link - multiple uses');
+          }
```

Comment:
Can't `user.lastLoginAt` be `null`? (it would crash in the condition below)

Reply:
You're right, when the db initialises, then `lastLoginAt` will be null. But it shouldn't crash in the `if` statement because of the code that decides whether you go into the `if` statement or not (line 104).

If it's the user's first login, i.e. they just signed up, then line 104 takes them around the `if` statement and straight into the bit where we overwrite the `lastLoginAt` in the db with the first login time. (This is because both of the conditions are falsy.)

So in all other cases, `user.lastloginAt` should actually be a timestamp, meaning we can: 1) prevent re-use of the first login token if `req.jwtPayload.lastLoginAt !== 'firstLogin'` evaluates to true; and 2) takes us into the `if` statement if it isn't the user's first login, allowing us to check that the `lastLoginAt` in the JWT matches that in the db.

Reply:
Alright, thanks for the details